src/smt/difficulty_post_processor.cpp

```diff
@@ -0,0 +1,77 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Implementation of module for processing the difficulty of input assumptions
+ * based on proof nodes.
+ */
+
+#include "smt/difficulty_post_processor.h"
+
+#include "smt/env.h"
+#include "util/rational.h"
+
+using namespace cvc5::kind;
+using namespace cvc5::theory;
+
+namespace cvc5 {
+namespace smt {
+
+DifficultyPostprocessCallback::DifficultyPostprocessCallback(Env& env)
+    : d_env(env), d_currDifficulty(0)
+{
+}
+
+bool DifficultyPostprocessCallback::setCurrentDifficulty(Node d)
+{
+  if (d.isConst() && d.getType().isInteger()
+      && d.getConst<Rational>().sgn() >= 0
+      && d.getConst<Rational>().getNumerator().fitsUnsignedInt())
+  {
+    d_currDifficulty = d.getConst<Rational>().getNumerator().toUnsignedInt();
+    return true;
+  }
+  return false;
+}
+
+bool DifficultyPostprocessCallback::shouldUpdate(std::shared_ptr<ProofNode> pn,
+                                                 const std::vector<Node>& fa,
+                                                 bool& continueUpdate)
+{
+  PfRule r = pn->getRule();
+  if (r == PfRule::ASSUME)
+  {
+    Trace("difficulty-debug")
+        << "  found assume: " << pn->getResult() << std::endl;
+    d_accMap[pn->getResult()] += d_currDifficulty;
```

Comment:
Seems really ugly to do this when checking whether to update. Couldn't you just move this block to an update method that does nothing else?

Reply:
This is also unclear since there is no updates.
Note the final proof callback also has this style: https://github.com/cvc5/cvc5/blob/master/src/smt/proof_final_callback.cpp#L56
Maybe `shouldUpdate` needs a more general name.

Reply:
Indeed. When I did something like that for letifying a proof node I had a similar situation but with `update` being a no-op for the proof node that updated internal information. 

I like `shouldUpdate` as is, but what we want is for updating only the internal state of the callback, which feels like a new method to be called in the pipeline between `shouldUpdate` and `update`.  What do you think?

Reply:
I'm actually just wondering if we should have a ProofVisitor class, which would be a simpler interface.

Reply:
Yes! That's the best solution. We already have three use cases, thus it's more than justified.

Reply:
Ok, let's do this in a later PR.